### PR TITLE
Detect extensions for test targets

### PR DESCRIPTION
**Pull request by kenzieschmoll**
Picking up work from https://github.com/flutter/devtools/pull/7206, which was closed due to merge conflicts and some changes that have since become obsolete.

Fixes https://github.com/flutter/devtools/issues/7183

### CODE REVIEW

File: packages/devtools_app/lib/src/service/service_manager.dart [hunk 1]
```diff
@@ -180,22 +175,82 @@ class ServiceConnectionManager {
     await serviceManager.isolateManager.init(isolates);
   }
 
+  /// Returns the package root URI for the connected app.
+  ///
+  /// This should be the directory up the tree from the debugging target that
+  /// contains the .dart_tool/package_config.json file.
+  ///
+  /// This method contains special logic for detecting the package root for
+  /// test targets (i.e. a VM service connections spawned from `dart test` or
```
Comment (bkonyi) [inline packages/devtools_app/lib/src/service/service_manager.dart:184]: The uberist of nits: `i.e.,`

File: packages/devtools_app/lib/src/service/service_manager.dart [hunk 2]
```diff
@@ -180,22 +175,82 @@ class ServiceConnectionManager {
...
+  Future<Library?> _mainIsolateRootLibrary() async {
+    final ref = (await serviceManager.isolateManager.waitForMainIsolateState())
+        ?.isolateNow
+        ?.rootLib;
+    if (ref == null) return null;
+    final library = await serviceManager.service!.getObject(
+      serviceManager.isolateManager.mainIsolate.value!.id!,
+      ref.id!,
+    );
+    return library is Library ? library : null;
```
Comment (bkonyi) [inline packages/devtools_app/lib/src/service/service_manager.dart:234]: Should we `assert(library is Library)` and then `return library as Library` instead? This should never resolve to anything but a `Library`.

**Review by bkonyi — APPROVED** [review]

File: packages/devtools_app/lib/src/service/service_manager.dart (hunk 2 above)
Comment (kenzieschmoll) [reply packages/devtools_app/lib/src/service/service_manager.dart:234]: couldn't this technically be a Sentinel?

Comment (kenzieschmoll) [reply packages/devtools_app/lib/src/service/service_manager.dart]: added a try / catch in case a SentinelException is thrown. PTAL.

Comment (bkonyi) [reply packages/devtools_app/lib/src/service/service_manager.dart]: Yep, that's the right thing to do. LGTM.

**Pull request merged**